opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py

```diff
@@ -114,9 +128,113 @@ def on_end(self, span: "Span") -> None:
             sp.on_end(span)
 
     def shutdown(self) -> None:
+        """Sequentially shuts down all underlying :class:`SpanProcessor`s
+        """
         for sp in self._span_processors:
             sp.shutdown()
 
+    def force_flush(self, timeout_millis: int = 30000) -> bool:
+        """Sequentially calls force_flush on all underlying
+        :class:`SpanProcessor`s
+
+        Args:
+            timeout_millis: The maximum amount of time over all SpanProcessor`s
+                to wait for spans to be exported. In case the first n span
+                processors exceeded the timeout followup span processors will be
+                skipped.
+
+        Returns:
+            True if all :class:`SpanProcessor`s flushed their spans within the
+            given timeout, False otherwise.
+        """
+        deadline_ns = time_ns() + timeout_millis * 1000000
+        for sp in self._span_processors:
+            current_time_ns = time_ns()
+            if current_time_ns >= deadline_ns:
+                return False
+
+            if not sp.force_flush((deadline_ns - current_time_ns) // 1000000):
+                return False
+
+        return True
+
+
+class ConcurrentMultiSpanProcessor(SpanProcessor):
+    """Implementation of :class:`SpanProcessor` that forwards all received
+    events to a list of `SpanProcessor`s in parallel.
+
+    Calls to the underlying `SpanProcessor`s are forwarded in parallel by
+    submitting them to a thread pool executor and waiting until each span
+    processor finished its work.
+
+    Args:
+        num_threads: The number of threads managed by the thread pool executor
+            and thus defining how many span processors can work in parallel.
+    """
+
+    def __init__(self, num_threads: int = 2):
+        # use a tuple to avoid race conditions when adding a new span and
```

Comment:
i took this one from the (Synchronous)`MultiSpanProcessor`.  in think in CPython a list is thread safe due to the global interpreter lock but might not be the case for other python interpreter implementations.

Reply:
The idea behind the initial implementation was to avoid using locks in `on_start` and `on_end` as these are called for every span in the system. I based my decision using the following considerations, some of them could be wrong:

- Appending to list is not thread safe in all the implementations (it's in CPython because of the GIL).
- Variable assignment is **atomic** in all implementations.

I wasn't able to find any authoritative answer about that, many of the documentation I found around is CPython specific, and AFAIU the Python specification doesn't state anything about operations atomicity, so the final decision is in the Python implementation.

It'd be very nice if somebody is able to clarify all these details to understand if the current implementation is 100% safe or not.